src/treeview/writer.js

```diff
@@ -833,9 +856,38 @@ function unwrapAttributeElement( wrapper, toUnwrap ) {
 // (`start` and `end` positions are located inside same {@link engine.treeView.AttributeElement AttributeElement}),
 // starts on 0 offset and ends after last child node.
 //
+// @private
 // @param {engine.treeView.Range} Range
 // @returns {Boolean}
 function rangeSpansOnAllChildren( range ) {
 	return range.start.parent == range.end.parent && range.start.parent instanceof AttributeElement &&
 		range.start.offset === 0 && range.end.offset === range.start.parent.getChildCount();
+}
+
+// Checks if provided nodes are valid to insert by writer. Checks if each node is an instance of
+// {@link engine.treeView.Text Text} or {@link engine.treeView.AttributeElement AttributeElement} or
+// {@link engine.treeView.ContainerElement ContainerElement}.
+//
+// Throws {@link utils.CKEditorError CKEditorError} `treeview-writer-insert-invalid-node` when nodes to insert
+// contains instances that are not {@link engine.treeView.Text Texts},
+// {@link engine.treeView.AttributeElement AttributeElements} or
+// {@link engine.treeView.ContainerElement ContainerElements}.
+//
+// @private
+// @param Iterable.<engine.treeView.Text|engine.treeView.AttributeElement|engine.treeView.ContainerElement> nodes
+function validateNodesToInsert( nodes ) {
+	for ( let node of nodes ) {
```

Comment:
The reason why I mentioned this issue is that the `getChildren()` call would be fine – we could keep that and everything would work. It's a thing that I don't like about iterables vs arrays... in one place you can't pass the object, because some other function would consume it once and for all, in other place it's just fine to make that simple call, without special conversion to an array or pass the array that you already have. And we could say that it's a theoretical issue, but as we've seen it's not. When writing a function like `validateNodesToInsert` you don't know how it will be used and in case of the iterables it's the user who needs to think about it. It's the opposite of the usual problem with arrays, which you need to remember to not touch inside that function... which is easier to do.

I guess I'll start an issue in the design repo tomorrow. I feel that this will be a recurring issue. Although perhaps the only thing which we need to say is that if a function accepts an iterable it should (or can) consume it and you should remember about this.